common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,905 @@
+       * going to write some data to disk, thus increasing the length counter. The other duplicate
+       * block is going to see the currentMapIndex already set to its mapIndex. However, it hasn't
+       * written any data yet. If the first block gets written completely and resets the
+       * currentMapIndex to -1 before the processing for the second block finishes, we can just
+       * check the bitmap to identify the second as a duplicate.
+       */
+      private boolean isDuplicateBlock() {
+        assert partitionInfo != null;
+        return (partitionInfo.getCurrentMapIndex() == mapIndex && length == 0)
+          || partitionInfo.mapTracker.contains(mapIndex);
+      }
+
+      /**
+       * This is only invoked when the stream is able to write. The stream first writes any deferred
+       * block parts buffered in memory.
+       */
+      private void writeAnyDeferredBufs() throws IOException {
+        assert partitionInfo != null;
+        if (deferredBufs != null && !deferredBufs.isEmpty()) {
+          for (ByteBuffer deferredBuf : deferredBufs) {
+            writeBuf(deferredBuf);
+          }
+          deferredBufs = null;
+        }
+      }
+
+      @Override
+      public void onData(String streamId, ByteBuffer buf) throws IOException {
+        // If partition info is null, ignore the requests. It could only be
+        // null either when a request is received after the shuffle file is
+        // finalized or when a request is for a duplicate block.
+        if (partitionInfo == null) {
+          return;
+        }
+        // When handling the block data using StreamInterceptor, it can help to reduce the amount
+        // of data that needs to be buffered in memory since it does not wait till the completion
+        // of the frame before handling the message, thus releasing the ByteBuf earlier. However,
+        // this also means it would chunk a block into multiple buffers. Here, we want to preserve
+        // the benefit of handling the block data using StreamInterceptor as much as possible while
+        // providing the guarantee that one block would be continuously written to the merged
+        // shuffle file before the next block starts. For each shuffle partition, we would track
+        // the current map index to make sure only block matching the map index can be written to
+        // disk. If one server thread sees the block being handled is the current block, it would
+        // directly write the block to disk. Otherwise, it would buffer the block chunks in memory.
+        // If the block becomes the current block before we see the end of it, we would then dump
+        // all buffered block data to disk and write the remaining portions of the block directly
+        // to disk as well. This way, we avoid having to buffer the entirety of every blocks in
+        // memory, while still providing the necessary guarantee.
```

Comment:
What is the worst case here in terms of aggregate memory used to buffer in-memory until it is written out ?

Reply:
This would be the max concurrent block streams received by the server that are experiencing collisions.
It shouldn't be a lot for the following reason:
To have a large number of concurrent block streams experiencing collisions, the number of shuffle partitions for a given shuffle should be large. However, if that's the case, the randomization on the client side before pushing the blocks would significantly reduce the likelihood of collision in the first place.

Reply:
But this will be across all applications here, right ?
I am trying to get a sense of what the potential impact on memory usage here would be (since it is outside of container restrictions like an executor - and given how lazy java is in releasing memory once acquired).

Reply:
So the factors to consider are the followings:
1. Number of concurrent shuffles handled by a given shuffle service for block merge
2. Number of shuffle partitions in each of these shuffles
3. likelihood of block collision for each of these shuffles
4. size of the shuffle blocks for these shuffles

We are effectively looking at 1*2*3*4 as the worse case scenario for the memory footprint.
2*3 should be small (10s for concurrent colliding blocks), due to the randomization on the client side.
4 also shouldn't be large (<1MB, average 10-100KB), otherwise we skip pushing these blocks in the first place.
1 is the only factor here that could be large in a busy cluster.
However, combined together, 1 needs to be extremely large to bring the memory footprint to 100sMB - GB level.
Also, since the block is cached in memory only for the duration of the stream, they are not kept in memory for long.

Reply:
Sounds good, this makes sense.
Given (1), do we want to enforce an upper bound limit to how many pending blocks can be kept in memory ?
This does not need to be in current PR, but could be in a follow up work - a jira could track this for later.

Thoughts @tgravescs ?

Reply:
@mridulm @tgravescs @Ngone51 I have create a follow-up jira for this:
https://issues.apache.org/jira/browse/SPARK-33331